addons/payment_demo/models/payment_transaction.py

```diff
@@ -132,16 +136,28 @@ def _get_tx_from_notification_data(self, provider_code, notification_data):
         :rtype: recordset of `payment.transaction`
         :raise: ValidationError if the data match no transaction
         """
-        tx = super()._get_tx_from_notification_data(provider_code, notification_data)
-        if provider_code != 'demo' or len(tx) == 1:
-            return tx
+        source_tx = super()._get_tx_from_notification_data(provider_code, notification_data)
+        if provider_code != 'demo' or len(source_tx) == 1:
+            return source_tx
 
         reference = notification_data.get('reference')
-        tx = self.search([('reference', '=', reference), ('provider_code', '=', 'demo')])
-        if not tx:
+        source_tx = self.search([('reference', '=', reference), ('provider_code', '=', 'demo')])
+
+        if not source_tx:
             raise ValidationError(
                 "Demo: " + _("No transaction found matching reference %s.", reference)
             )
+        else:
+            notification_data_amount = notification_data.get('amount', '')
+            if source_tx.amount != notification_data_amount:  # partial
+                tx = self.search([
+                    ('provider_reference', '=', reference),
+                    ('provider_code', '=', 'demo'),
+                ])
+            else:
+                tx = source_tx
+            if not tx:
+                tx = source_tx._create_child_transaction(notification_data_amount)
```

Comment:
Same here, if a child tx was needed, it was created long before you process the notification data =]